src/version_req.rs

```diff
@@ -407,6 +408,315 @@ impl VersionReq {
 
         false
     }
+
+    /// Return a `VersionReq` that matches any version that matches _either_ the current
+    /// requirements or those expressed in `other`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `self` and `other` have different compatibilities.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use semver::VersionReq;
+    /// use semver::Version;
+    ///
+    /// let version_110 = Version { major: 1, minor: 1, patch: 0, pre: vec![], build: vec![] };
+    /// let version_111 = Version { major: 1, minor: 1, patch: 1, pre: vec![], build: vec![] };
+    /// let match_110 = VersionReq::parse("=1.1.0").unwrap();
+    /// let match_111 = VersionReq::parse(">1.1.0").unwrap();
+    /// let either = match_110.union(&match_111);
+    ///
+    /// assert!(either.matches(&version_110));
+    /// assert!(either.matches(&version_111));
+    /// ```
+    pub fn union(mut self, other: &VersionReq) -> VersionReq {
+        assert_eq!(self.compat, other.compat);
+        self.ranges.extend(other.ranges.iter().cloned());
+        self.simplify();
+        self
+    }
+
+    /// Return a `VersionReq` that matches any version that matches the current requirements _and_
+    /// those expressed in `other`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `self` and `other` have different compatibilities.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use semver::VersionReq;
+    /// use semver::Version;
+    ///
+    /// let version_110 = Version { major: 1, minor: 1, patch: 0, pre: vec![], build: vec![] };
+    /// let version_111 = Version { major: 1, minor: 1, patch: 1, pre: vec![], build: vec![] };
+    /// let version_120 = Version { major: 1, minor: 2, patch: 0, pre: vec![], build: vec![] };
+    /// let gt_110 = VersionReq::parse(">1.1.0").unwrap();
+    /// let lt_120 = VersionReq::parse("<1.2.0").unwrap();
+    /// let both = gt_110.intersection(&lt_120);
+    ///
+    /// assert!(both.matches(&version_111));
+    /// assert!(!both.matches(&version_110));
+    /// assert!(!both.matches(&version_120));
+    /// ```
+    pub fn intersection(mut self, other: &VersionReq) -> VersionReq {
```

Comment:
`mut self` ... `VersionReq` is an odd interface. Is it modifying inplace or returning a new one?

Reply:
`VersionReq` is not `Copy`, so this consumes `self` and returns a modified version. We could do an internal variable assignment to have it not say `mut self`, but not sure if that's worth it? Is the `mut` part even exposed in the docs given that it's more of a consumer detail?

The reason it's not `&self` is that it would have to allocate a new `VersionReq` for the result, which seems unfortunate. Passing in ownership means that the caller can choose to allocate _if they wish_ by calling `clone` themselves.

The reason it's not `&mut self` is that (I think) it's less ergonomic.You'd have to write:
```rust
vreq.union(&other);
if vreq.matches(&v) {
```
as opposed to
```rust
if vreq.union(&other).matches(&v) {
```

Reply:
You are entirely correct. I was miss reading it as `&mut self`.